lib/msprime.c

```diff
@@ -4223,6 +4222,82 @@ msp_add_instantaneous_bottleneck(msp_t *self, double time, int population_id,
     return ret;
 }
 
+/* Add a census event at a specified time (given in generations). */
+static int
+msp_census_event(msp_t *self, demographic_event_t *event)
+{
+    int ret = 0;
+    double time = self->model.model_time_to_generations(&self->model, event->time);
+    avl_tree_t *ancestors;
+    avl_node_t *node;
+    segment_t *seg;
+    int i, j;
+
+    for (i = 0; i < (int) self->num_populations; i++) {
+        for (j = 0; j < (int) self->num_labels; j++) {
+
+            // Get segment from an ancestor in a population.
+            ancestors = &self->populations[i].ancestors[j];
+            node = ancestors->head;
+
+            while (node != NULL){
+                seg = node->item;
+
+                // Add an edge to the edge table.
+                ret = msp_flush_edges(self);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_node_table_add_row(&self->tables->nodes, 0, time, 0,
+                        TSK_NULL, NULL, 0);
+
+                // Add an edge joining the segment to the new node.
+                ret = msp_store_edge(self, seg->left, seg->right, ret, seg->label);
+
+                // Modify segment node id.
+                seg->label = ret;
+                ret = 0;
+                node = node->next;
+            }
+        }
+    }
+
+out:
+    return ret;
+}
+
+static void
+msp_print_census_event(msp_t * MSP_UNUSED(self), demographic_event_t *event, FILE *out)
+{
+    fprintf(out, "%f\tcensus_event:\n",
+            event->time);
+}
+
+int MSP_WARN_UNUSED
+msp_add_census_event(msp_t *self, double time)
+{
+    int ret = 0;
+    demographic_event_t *de;
+
+    if (time <= 0) {
```

Comment:
Is this caugnt in ``add_demographic_event``? If so, we don't need it.

Reply:
The only check of time I can see in `msp_add_demographic_event` is [this one](https://github.com/tskit-dev/msprime/blob/master/lib/msprime.c#L3635-L3640), which is about sortedness rather than times being < 0:

```
    if (self->demographic_events_tail != NULL) {
        if (time < self->demographic_events_tail->time) {
            ret = MSP_ERR_UNSORTED_DEMOGRAPHIC_EVENTS;
            goto out;
        }
    }
```
so maybe it would be better to move this check of time > 0 into `add_demographic_event` instead?

Reply:
Yes, maybe. Try it out and see if it breaks stuff. If it does then forget about it.

Reply:
I've put a check of time < 0 into `add_demographic_event` (time == 0 breaks it).

Perhaps I could also put in a check that there aren't any existing nodes at the census time for `add_census_event`? This is a more general test that would also catch the time == 0 case. I think you said that some there is already some error that might be thrown if the census time is identical to a coalescence time, but perhaps it would be useful to have a more specific/interpretable error message for this situation?

Reply:
NB. esp for the DTWF situation where users might be inclined to set a census at an integer time.